src/utils.rs

```diff
@@ -8,13 +8,11 @@ pub fn time() -> u128 {
 pub fn extract_id_from_filename(
     entry: &std::path::PathBuf,
 ) -> Result<u128, Box<dyn std::error::Error>> {
-    let id = entry
+    entry
         .extension()
-        .ok_or(string_error::new_err(
-            "Missing extension (ie. not in format: data.<id>)",
-        ))?
+        .ok_or_else(|| string_error::new_err("Missing extension (ie. not in format: data.<id>)"))?
```

Comment:
this change is an important thing to learn

rust has two standard forms for combinators:
**immediately evaluated**, these take the value directly. such as
* `ok_or`
* `unwrap_or`


**lazily evaluated**, these generally end with `_else`. they take a `FnOnce()` that produce the value lazily. such as
* `ok_or_else`
* `unwrap_or_else`


the immediate ones _always_ evaluate their arguments even if that "path" isn't take. so the prior code was constructing that string error even if the `extension` returned `Some`

the lazy ones take a closure which produces the new value.
this closure is only called when that "path" is taken.